Microsoft/helium
# Support case-insensitive Movie Title search

Bart added a `textSearch` column to the Movie documents that looks like this:

>  "textSearch": "the pianist genre:biography genre:drama genre:music adrien brody character:wladyslaw szpilman thomas kretschmann character:captain wilm hosenfeld frank finlay character:father emilia fox character:dorota roman polanski robert benmussa alain sarde"

Since the case will always be lowercase, we can convert the movie title passed in by the caller to lowercase and use `StartsWith` to find the right movie. `textSearch` will always begin with the lowercase version of the movie title.

`StartsWith` takes advantage of the CosmosDB Range Index while using `Lower` in the query doesn't use any index and causes a full DB scan, hence why I have lowercasing handled in the server code and not the DB query.

**Note:** Specifically for this hack, its unknown how the movie titles were converted to lowercase, and whether we will hit an culture issues (ex: toLowerCase with Turkish doesn't work like we might expect).

To test this change:

1. Checkout the changes in this PR (see link at bottom of PR)
2. Build and run the code following the steps in the standard Helium README
3. Run the following two commands:

```
curl -G 'localhost:3000/api/movies' --data-urlencode 'q=The Pianist' | jq
curl -G 'localhost:3000/api/movies' --data-urlencode 'q=tHe pIAnist' | jq
```

4. Rejoice when both commands return a movie with 'The Pianist' as the title.
